File: CalculoDespegueCessna172N.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class CalculoDespegueCessna172N {

    private static final Map<String, Map<String, int[]>> tablaCombinada = new HashMap<>();

    static {
        Map<String, int[]> tabla0C = new HashMap<>();
        tabla0C.put("PRESS_ALT_FT", new int[]{0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000});
        tabla0C.put("TOTAL_TO_CLEAR_50_FT_OBS", new int[]{1250, 1370, 1510, 1675, 1859, 2061, 2314, 2611, 2975});
        tablaCombinada.put("0C", tabla0C);

        Map<String, int[]> tabla10C = new HashMap<>();
        tabla10C.put("PRESS_ALT_FT", new int[]{0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000});
        tabla10C.put("TOTAL_TO_CLEAR_50_FT_OBS", new int[]{1340, 1475, 1625, 1800, 2000, 2235, 2510, 2850, 3270});
        tablaCombinada.put("10C", tabla10C);

        Map<String, int[]> tabla20C = new HashMap<>();
        tabla20C.put("PRESS_ALT_FT", new int[]{0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000});
        tabla20C.put("TOTAL_TO_CLEAR_50_FT_OBS", new int[]{1440, 1585, 1750, 1940, 2165, 2425, 2740, 3125, 3610});
        tablaCombinada.put("20C", tabla20C);

        Map<String, int[]> tabla30C = new HashMap<>();
        tabla30C.put("PRESS_ALT_FT", new int[]{0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000});
        tabla30C.put("TOTAL_TO_CLEAR_50_FT_OBS", new int[]{1545, 1705, 1885, 2095, 2340, 2635, 2985, 3430, 4000});
        tablaCombinada.put("30C", tabla30C);

        Map<String, int[]> tabla40C = new HashMap<>();
        tabla40C.put("PRESS_ALT_FT", new int[]{0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000});
        tabla40C.put("TOTAL_TO_CLEAR_50_FT_OBS", new int[]{1655, 1830, 2030, 2260, 2535, 2860, 3265, 3775, 4465});
        tablaCombinada.put("40C", tabla40C);
    }

    public static Map<String, Integer> obtenerDatosAltitud(int altitud, int temperatura) {
        String tablaTemperatura;
        if (temperatura < 0 || temperatura > 40) {
            System.out.println("Temperatura fuera de rango (0°C - 40°C).");
            return null;
        } else if (temperatura < 10) {
            tablaTemperatura = "0C";
        } else if (temperatura < 20) {
            tablaTemperatura = "10C";
        } else if (temperatura < 30) {
            tablaTemperatura = "20C";
        } else if (temperatura < 40) {
            tablaTemperatura = "30C";
        } else {
            tablaTemperatura = "40C";
        }

        Map<String, int[]> tablaSeleccionada = tablaCombinada.get(tablaTemperatura);
        for (int i = 0; i < tablaSeleccionada.get("PRESS_ALT_FT").length; i++) {
            if (tablaSeleccionada.get("PRESS_ALT_FT")[i] == altitud) {
                return Map.of("PRESS ALT FT", altitud, "TOTAL TO CLEAR 50 FT OBS", tablaSeleccionada.get("TOTAL_TO_CLEAR_50_FT_OBS")[i]);
            }
        }
        System.out.println("No se encontraron datos de altitud para la altitud y temperatura seleccionadas.");
        return null;
    }

    public static void calcularPerformanceDistancia(int temperatura, int vientoContra, int vientoCola, String tipoPista, String aeropuerto) {
        Map<String, Double> elevacionesPreestablecidas = Map.of(
                "mgpb", 32.80,
                "mgsj", 45.93,
                "mgrd", 65.61,
                "mgrb", 465.88,
                "mgza", 633.20,
                "mgrt", 656.16,
                "mgpp", 1692.91,
                "mggt", 5000.0,
                "mgqz", 7808.40,
                "mgsm", 7939.63,
                "mgqc", 6614.17
        );
        String aeropuertoLower = aeropuerto.toLowerCase();
        double elevacionPies;
        if (elevacionesPreestablecidas.containsKey(aeropuertoLower)) {
            elevacionPies = elevacionesPreestablecidas.get(aeropuertoLower);
        } else {
            Scanner scanner = new Scanner(System.in);
            System.out.print("Ingresa la altura del aeropuerto (en pies): ");
            elevacionPies = scanner.nextDouble();
        }

        double xPorcentaje = vientoContra / 9.0;
        Map<String, Integer> datosAltitud = obtenerDatosAltitud((int) elevacionPies, temperatura);
        if (datosAltitud != null) {
            int totalToClear = datosAltitud.get("TOTAL TO CLEAR 50 FT OBS");

            double distanciaNecesariaPies = totalToClear - (totalToClear * (xPorcentaje / 100.0));

            if (vientoCola <= 10) {
                distanciaNecesariaPies *= 1.10;
            }
            if (tipoPista.equalsIgnoreCase("dry") || tipoPista.equalsIgnoreCase("grass runway")) {
                distanciaNecesariaPies *= 1.15;
            }

            System.out.println("Datos de Altitud: " + datosAltitud);
            System.out.println("Distancia necesaria para despegue (pies): " + distanciaNecesariaPies);
        } else {
            System.out.println("No se pudieron calcular los datos de altitud para la altitud y temperatura seleccionadas.");
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        while (true) {
            System.out.println("\nBienvenido al Calculador de Distancia para el Despegue de un Cessna 172N\n"
                    + "Este sistema te permite calcular la distancia necesaria para el despegue de un Cessna 172N "
                    + "considerando varias condiciones, como longitud de la pista, temperatura del aeropuerto de salida, "
                    + "viento y tipo de pista. Este programa está diseñado para vuelos internos en Guatemala, o para"
                    + "salida de vuelos internacionales partiendo de aeropuertos guatemaltecos.\n");

            System.out.println("\nLista de aeropuertos/aeródromos de salida:");
            System.out.println("| LUGAR             | INDICADOR |");
            System.out.println("|-------------------|-----------|");
            System.out.println("| PUERTO BARRIOS    | MGPB      |");
            System.out.println("| PUERTO DE SAN JOSÉ| MGSJ      |");
            System.out.println("| BARILLAS          | MGBA      |");
            System.out.println("| BANANERA          | MGBN      |");
            System.out.println("| COBÁN             | MGCB      |");
            System.out.println("| CHAHAL            | MGCH      |");
            System.out.println("| CHINAJÁ           | MGCJ      |");
            System.out.println("| CHAMÁ             | MGCM      |");
            System.out.println("| CHAMPERICO        | MGCP      |");
            System.out.println("| CHIQUIMULA        | MGCQ      |");
            System.out.println("| CARMELITA         | MGCR      |");
            System.out.println("| CHISEC            | MGCS      |");
            System.out.println("| COATEPEQUE        | MGCT      |");
            System.out.println("| DOS LAGUNAS       | MGDL      |");
            System.out.println("| ESQUIPULAS        | MGES      |");
            System.out.println("| FRAY BARTOLOMÉ    |           |");
            System.out.println("| DE LAS CASAS      | MGFB      |");
            System.out.println("| LA AURORA         | MGGT      |");
            System.out.println("| HUEHUETENANGO     | MGHT      |");
            System.out.println("| SAN JERÓNIMO      | MGJE      |");
            System.out.println("| JUTIAPA           | MGJU      |");
            System.out.println("| LA LIBERTAD       | MGLL      |");
            System.out.println("| MUNDO MAYA        | MGMM      |");
            System.out.println("| PLAYA GRANDE      | MGPG      |");
            System.out.println("| PETÉN ITZÁ        | MGPI      |");
            System.out.println("| POPTÚN            | MGPP      |");
            System.out.println("| QUICHÉ            | MGQC      |");
            System.out.println("| QUETZALTENANGO    | MGQZ      |");
            System.out.println("| RABINAL           | MGRA      |");
            System.out.println("| RUBELSANTO        | MGRB      |");
            System.out.println("| RÍO DULCE         | MGRD      |");
            System.out.println("| RESURRECCIÓN      | MGRE      |");
            System.out.println("| RETALHULEU        | MGRT      |");
            System.out.println("| SAN ANDRÉS        |           |");
            System.out.println("| SAJCABAJÁ         | MGSA      |");
            System.out.println("| SAN MARCOS        | MGSM      |");
            System.out.println("| SAYAXCHÉ          | MGSX      |");
            System.out.println("| TILAPA            | MGTI      |");
            System.out.println("| USPANTÁN          | MGUS      |");
            System.out.println("| XALBAL            | MGXB      |");
            System.out.println("| ZACAPA            | MGZA      |\n");

            System.out.print("Por favor, ingresa el indicador del aeropuerto/aeródromo de salida: ");
            String aeropuertoSalida = scanner.nextLine();
            System.out.print("Por favor, ingresa la temperatura del aeropuerto de salida (en grados Celsius): ");
            int temperatura = scanner.nextInt();
            System.out.print("¿Tienes viento en contra? (nudos): ");
            int vientoContra = scanner.nextInt();
            double vientoCola;
            if (vientoContra == 0) {
                System.out.print("¿Cuántos nudos de viento en cola tienes? ");
                vientoCola = scanner.nextDouble();
            } else {
                System.out.print("¿Cuántos nudos de viento en cola tienes? ");
                vientoCola = scanner.nextDouble();
            }
            scanner.nextLine(); // Limpiar buffer
            System.out.print("Por favor, ingresa el tipo de pista (dry, grass, normal): ");
            String tipoPista = scanner.nextLine();

            calcularPerformanceDistancia(temperatura, vientoContra, (int) vientoCola, tipoPista, aeropuertoSalida);

            System.out.print("¿Deseas realizar otro cálculo? (s/n): ");
            String continuar = scanner.nextLine();
            if (!continuar.equalsIgnoreCase("s")) {
                System.out.println("¡Gracias por usar el Calculador de Distancia para el Despegue!");
                break;
            }
        }
    }
}

File: Calculo Despegue Cessna 172N.py
def obtener_datos_altitud(altitud, temperatura):
    tabla_combinada = {
        "0C": {
            "PRESS_ALT_FT": [0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000],
            "TOTAL_TO_CLEAR_50_FT_OBS": [1300, 1420, 1555, 1710, 1880, 2075, 2305, 2565, 2870]
        },
        "10C": {
            "PRESS_ALT_FT": [0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000],
            "TOTAL_TO_CLEAR_50_FT_OBS": [1390, 1525, 1670, 1835, 2025, 2240, 2485, 2770, 3110]
        },
        "20C": {
            "PRESS_ALT_FT": [0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000],
            "TOTAL_TO_CLEAR_50_FT_OBS": [1490, 1630, 1790, 1970, 2175, 2410, 2680, 3000, 3375]
        },
        "30C": {
            "PRESS_ALT_FT": [0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000],
            "TOTAL_TO_CLEAR_50_FT_OBS": [1590, 1745, 1915, 2115, 2335, 2595, 2895, 3245, 3670]
        },
        "40C": {
            "PRESS_ALT_FT": [0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000],
            "TOTAL_TO_CLEAR_50_FT_OBS": [1700,]
        }
    }
    if temperatura < 0 or temperatura > 40:
        print("Temperatura fuera de rango (0°C - 40°C).")
        return None
    elif temperatura < 10:
        tabla_temperatura = "0C"
    elif temperatura < 20:
        tabla_temperatura = "10C"
    elif temperatura < 30:
        tabla_temperatura = "20C"
    elif temperatura < 40:
        tabla_temperatura = "30C"
    else:
        tabla_temperatura = "40C"  # Corregido para manejar temperatura igual a 40°C

    tabla_seleccionada = tabla_combinada[tabla_temperatura]
    for i in range(len(tabla_seleccionada["PRESS_ALT_FT"])):
        if tabla_seleccionada["PRESS_ALT_FT"][i] == altitud:
            return {"PRESS ALT FT": altitud,
                    "TOTAL TO CLEAR 50 FT OBS": tabla_seleccionada["TOTAL_TO_CLEAR_50_FT_OBS"][i]}
    print("No se encontraron datos de altitud para la altitud y temperatura seleccionadas.")
    return None


def calcular_performance_distancia(temperatura, viento_contra, viento_cola, tipo_pista, aeropuerto):
    elevaciones_preestablecidas = {
        "MGPB": 32.80,
        "MGSJ": 45.93,
        "MGRD": 65.61,
        "MGRB": 465.88,
        "MGZA": 633.20,
        "MGRT": 656.16,
        "MGPP": 1692.91,
        "MGGT": 5000,
        "MGQZ": 7808.40,
        "MGSM": 7939.63,
        "MGQC": 6614.17
    }
    aeropuerto_upper = aeropuerto.upper()  # Convertido a mayúsculas para manejar correctamente los códigos de aeropuerto
    if aeropuerto_upper in elevaciones_preestablecidas:
        elevacion_pies = elevaciones_preestablecidas[aeropuerto_upper]
        x_porcentaje = viento_contra / 9
        datos_altitud = obtener_datos_altitud(elevacion_pies, temperatura)
        if datos_altitud:
            total_to_clear = datos_altitud["TOTAL TO CLEAR 50 FT OBS"]

            distancia_necesaria_pies = total_to_clear - (total_to_clear * (x_porcentaje / 100))

            if viento_cola <= 10:
                distancia_necesaria_pies *= 1.10
            if tipo_pista.lower() in ["dry", "grass runway"]:  # Simplificado el chequeo del tipo de pista
                distancia_necesaria_pies *= 1.15

            print("Datos de Altitud:", datos_altitud)
            print("Distancia necesaria para el despegue (pies):", distancia_necesaria_pies)
        else:
            print("No se pudieron calcular los datos de altitud para la altitud y temperatura seleccionadas.")
    else:
        elevacion_pies = float(input("Ingresa la altura del aeropuerto (en pies): "))
        x_porcentaje = viento_contra / 9
        datos_altitud = obtener_datos_altitud(elevacion_pies, temperatura)
        if datos_altitud:
            total_to_clear = datos_altitud["TOTAL TO CLEAR 50 FT OBS"]

            distancia_necesaria_pies = total_to_clear - (total_to_clear * (x_porcentaje / 100))

            if viento_cola <= 10:
                distancia_necesaria_pies *= 1.10
            if tipo_pista.lower() in ["dry", "grass runway"]:  # Simplificado el chequeo del tipo de pista
                distancia_necesaria_pies *= 1.15

            print("Datos de Altitud:", datos_altitud)
            print("Distancia necesaria para el despegue (pies):", distancia_necesaria_pies)
        else:
            print("No se pudieron calcular los datos de altitud para la altitud y temperatura seleccionadas.")


def main():
    while True:
        print("\nBienvenido al Calculador de Distancia para el Despegue de un Cessna 172N\n"
              "Este sistema te permite calcular la distancia necesaria para el despegue de un Cessna 172N "
              "considerando varias condiciones, como longitud de la pista, temperatura del aeropuerto de salida, "
              "viento y tipo de pista. Este programa está diseñado para vuelos internos en Guatemala, o para"
              "salida de vuelos internacionales partiendo de aeropuertos guatemaltecos.\n")

        print("\nLista de aeropuertos/aeródromos de salida:")
        print("| LUGAR             | INDICADOR |")
        print("|-------------------|-----------|")
        print("| PUERTO BARRIOS    | MGPB      |")
        print("| PUERTO DE SAN JOSÉ| MGSJ      |")
        print("| BARILLAS          | MGBA      |")
        print("| BANANERA          | MGBN      |")
        print("| COBÁN             | MGCB      |")
        print("| CHAHAL            | MGCH      |")
        print("| CHINAJÁ           | MGCJ      |")
        print("| CHAMÁ             | MGCM      |")
        print("| CHAMPERICO        | MGCP      |")
        print("| CHIQUIMULA        | MGCQ      |")
        print("| CARMELITA         | MGCR      |")
        print("| CHISEC            | MGCS      |")
        print("| COATEPEQUE        | MGCT      |")
        print("| DOS LAGUNAS       | MGDL      |")
        print("| ESQUIPULAS        | MGES      |")
        print("| FRAY BARTOLOMÉ    |           |")
        print("| DE LAS CASAS      | MGFB      |")
        print("| LA AURORA         | MGGT      |")
        print("| HUEHUETENANGO     | MGHT      |")
        print("| SAN JERÓNIMO      | MGJE      |")
        print("| JUTIAPA           | MGJU      |")
        print("| LA LIBERTAD       | MGLL      |")
        print("| MUNDO MAYA        | MGMM      |")
        print("| PLAYA GRANDE      | MGPG      |")
        print("| PETÉN ITZÁ        | MGPI      |")
        print("| POPTÚN            | MGPP      |")
        print("| QUICHÉ            | MGQC      |")
        print("| QUETZALTENANGO    | MGQZ      |")
        print("| RABINAL           | MGRA      |")
        print("| RUBELSANTO        | MGRB      |")
        print("| RÍO DULCE         | MGRD      |")
        print("| RESURRECCIÓN      | MGRE      |")
        print("| RETALHULEU        | MGRT      |")
        print("| SAN ANDRÉS        |           |")
        print("| SAJCABAJÁ         | MGSA      |")
        print("| SAN MARCOS        | MGSM      |")
        print("| SAYAXCHÉ          | MGSX      |")
        print("| TILAPA            | MGTI      |")
        print("| USPANTÁN          | MGUS      |")
        print("| XALBAL            | MGXB      |")
        print("| ZACAPA            | MGZA      |\n")

        aeropuerto_salida = input("Por favor, ingresa el indicador del aeropuerto/aeródromo de salida: ").upper()  # Convertido a mayúsculas
        temperatura = float(input("Por favor, ingresa la temperatura del aeropuerto de salida (en grados Celsius): "))
        viento_contra = float(input("¿Tienes viento en contra? (nudos): "))

        if viento_contra == 0:
            viento_cola = float(input("¿Cuántos nudos de viento en cola tienes? "))
        else:
            viento_cola = float(input("¿Cuántos nudos de viento en cola tienes? "))

        tipo_pista = input("Por favor, ingresa el tipo de pista (dry, grass, normal): ")

        calcular_performance_distancia(
            temperatura=temperatura,
            viento_contra=viento_contra,
            viento_cola=viento_cola,
            tipo_pista=tipo_pista,
            aeropuerto=aeropuerto_salida
        )

        continuar = input("¿Deseas realizar otro cálculo? (s/n): ")
        if continuar.lower() != "s":
            print("¡Gracias por usar el Calculador de Distancia para el Despegue!")
            break


if __name__ == "__main__":
    main()
